bson/bsonrw/json_scanner.go

```diff
@@ -202,13 +236,59 @@ func (js *jsonScanner) scanString() (*jsonToken, error) {
 					return nil, fmt.Errorf("invalid unicode sequence in JSON string: %s", us)
 				}
 
-				s := fmt.Sprintf(`\u%s`, us)
-				s, err = strconv.Unquote(strings.Replace(strconv.Quote(s), `\\u`, `\u`, 1))
-				if err != nil {
-					return nil, err
+				rn := getu4(us)
+
+				// If the rune we just decoded is the high or low value of a possible surrogate pair,
+				// try to decode the next sequence as the low value of a surrogate pair. We're expecting
+				// something like "\uDD1E", but need to handle cases where the input is not a valid
+				// surrogate pair. For more context on unicode surrogate pairs, see:
+				// https://www.christianfscott.com/rust-chars-vs-go-runes/
+				// https://www.unicode.org/glossary/#high_surrogate_code_point
+				if utf16.IsSurrogate(rn) {
+					c, err = js.readNextByte()
+					if err != nil {
+						return nil, errors.New("end of input in JSON string")
+					}
+
+					// If the next value isn't the beginning of a backslash escape sequence, write the
+					// current rune and goto the beginning of the next char eval block.
+					if c != '\\' {
+						b.WriteRune(rn)
+						goto evalNextChar
```

Comment:
Yes, this code uses `goto`s 🙁. Unfortunately, it's the cleanest way I've found so far to express this logic in the current code.

Other things I tried that seemed initially to be better but turned out worse:
- Write a `peekNNextBytes` that can look ahead in the `jsonScanner` N bytes without advancing the read position. This fails because the `jsonScanner` is set up with the expectation that the caller is always reading one byte at a time. If your peek causes a `readIntoBuf` that returns an error but you don't consume the bytes and error in the current block, subsequent re-reads of the same data will result in that error surfacing early, breaking parsing. Updating the `jsonScanner` code to support reading ahead many bytes is a significant refactor.

- Post-process the string/bytes in the `bytes.Buffer` filled by `jsonScanner.scanString` if we detect any possible surrogate pairs. This fails because `bytes.Buffer.WriteRune` encodes all `rune`s written as UTF-8 bytes. Unicode surrogate values are not individually valid UTF-8 code points and are converted to the Unicode replacement character instead when converted to UTF-8. The only way to retain the original `rune` value and post-process them is to keep the values in a `[]rune`, which is a significant refactor.